### ISSUE TITLE
Common UI component fails after migration from {N} 6.8 to 8.1 with error: instanceType is not a constructor

### ISSUE THREAD

**Issue by dlcole** [NativeScript/NativeScript#9691]
### Issue Description

I created a very simple common UI component, blue-text, and included it in a Hello World Nativescript 6.8 JavaScript project;  It works as expected.  

blue-text.js: 

```
/**
 * BlueText Custom UI Component
 * Derived from https://github.com/mikebranstein/TheNativeScriptBook/blob/master/AppendixB/PetScrapbook/app/views/shared/scrapbook-list/scrapbook-list.js 
 * Invocation: 
 * <page xmlns:bt="shared/blue-text/blue-text" ... />
 * <bt:BlueText text="This is blue text" ... />
 */

// For {N} 6
const Label = require("tns-core-modules/ui/label").Label;

// For {N} 8
// import {Label} from "@nativescript/core";

var BlueText = (function (_super) {

  __extends(BlueText, _super);

  function BlueText() {
    _super.call(this);

    this.style = "color: blue;"; 

  }; // end BlueText function
  return BlueText;
})(Label);  // Specify parent class

exports.BlueText = BlueText;
```
main-page.xml: 

```
<Page xmlns="http://schemas.nativescript.org/tns.xsd" xmlns:bt="shared/blue-text/blue-text" navigatingTo="onNavigatingTo">
...
<bt:BlueText text="This is blue text" />
...
```
Then, I upgrade the NativeScript CLI to 8.1.5, migrate the project, make the changes necessary eliminate the build errors (e.g., removing tildes, changing require's to import's, etc) and run the project.  It fails with the message 

```
System.err:  > Building UI from XML. @main-page.xml:30:9
System.err:    > Module 'shared/blue-text/blue-text' not found for element 'shared/blue-text/blue-text:BlueText'.
System.err:    > instanceType is not a constructor
```
This might be a Webpack problem or a problem caused by moving to ES2017, but I've been unable to find any information to make progress.  
 

### Reproduction

https://github.com/dlcole/cctester

### Relevant log output (if applicable)

```shell
System.err:  > Building UI from XML. @main-page.xml:30:9
System.err:    > Module 'shared/blue-text/blue-text' not found for element 'shared/blue-text/blue-text:BlueText'.
System.err:    > instanceType is not a constructor
```

### Environment

```yaml
OS: macOS 11.6
CPU: (8) x64 Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
Shell: /bin/bash
node: 16.13.0
npm: 8.1.0
nativescript: 8.1.5

# android
java: 1.8.0_192
ndk: Not Found
apis: 28
build_tools: 28.0.3
system_images: 
  - android-28 | Google APIs Intel x86 Atom

# ios
xcode: 13.1/13A1030d
cocoapods: 1.8.4
python: 2.7.16
python3: 3.9.5
ruby: 2.6.3
platforms: 
  - DriverKit 21.0.1
  - iOS 15.0
  - macOS 12.0
  - tvOS 15.0
  - watchOS 8.0
```

### Dependencies

```json
"dependencies": {
  "@nativescript/core": "~8.1.1",
  "@nativescript/theme": "~2.3.0"
},
"devDependencies": {
  "@nativescript/android": "8.1.1",
  "@nativescript/webpack": "~5.0.0"
}
```

### 

- [X] I have searched the [existing issues](https://github.com/NativeScript/NativeScript/issues) as well as [StackOverflow](https://stackoverflow.com/questions/tagged/nativescript) and this has not been posted before
- [X] This is a bug report
- [X] I agree to follow this project's [Code of Conduct](https://github.com/NativeScript/NativeScript/blob/master/tools/notes/CONTRIBUTING.md#coc)

**Comment by NathanWalker** [comment]
Try changing:
exports.BlueText
to:
export { BlueText };

**Comment by dlcole** [comment]
@NathanWalker - that at least gives me a different error message: 

```
> Building UI from XML. @main-page.xml:30:9
> Module 'shared/blue-text/blue-text' not found for element 'shared/blue-text/blue-text:BlueText'.
> Class constructor Label cannot be invoked without 'new'
```

I'm following the example code that was in version 1 of The NativeScript Book, as I haven't found any other documentation on the subject.  It's worked up until now, but I'd welcome any additional insights.

**Comment by NathanWalker** [comment]
No worries, you're running into esm vs commonjs stuff which is expected with {N} 6 > 7 and above.

Try this:

```
import {Label} from "@nativescript/core";

export class BlueText extends Label {
  initNativeView() {
    super.initNativeView();
    this.style = 'color: blue;';
  }
}
```

**Comment by dlcole** [comment]
Ha!  That works!  Thanks so much for your help and quick response, @NathanWalker 

I'll close this issue, but it can likely help anyone else trodding the same migration path.

**Comment by dlcole** [comment]
@NathanWalker - I stumped again on a related migration issue.  I have several common UI components I need to migrate.  So far, so good, EXCEPT for custom events defined for those components.  For example, I have a custom GridTab component that can look like this: 

`<gridtab:GridTab keys="{{ eventData.announcementsByDay }}" tabTap="onGridTabTap" />`

This creates a GridLayout with Labels from the keys property.  I have code that receives control when the label is tapped, namely: 

    ```
// Process tab tap event and bubble up to other handlers 
    var onTabTap = function (obj, onTapEventName) {

      // Move highlighting (via css) to selected tab
      let gridTab = obj.parent;
      gridTab.eachChild((child) => {
        child.class = "gridtab-tab"
      });
      obj.class = "gridtab-tab gridtab-selected";
      
      // Bubble up tabTap event to other handlers
      this.notify({ 
        eventName: onTapEventName,
        object: obj
      });
    }.bind(this);
```
This function is bound to each label and it receives control as expected.  But, the `notify()` has no effect.  From stepping though the execution it appears there is no global listener for the parent element GridTab.  In the V6 version of the code there is this line: 

`GridTab.tabTapEvent = "tabTap"; // Specify custom events for this class`

The NativeScript Book V1 describes this as a convention for identifying custom events, but this seems to have no effect now.  

My question:  in V8, how are custom events specified in a custom component?

**Comment by dlcole** [comment]
It took me two days to figure this out, but here's what I've found:  in NS 6, the line 

`GridTab.tabTapEvent = "tabTap"; // Specify custom events for this class`

apparently created an event listener, and that doesn't happen with NS 7 and later.  That means the custom component must create its own listener, such as (within onLoaded): 

```
    if (this.tabTap) {
      let handler = this.page.exports[this.tabTap]; // get associated handler 
      this.on("tabTap", handler, this); // add listener to receive notify
    }
```

Now, there's a listener in place to receive the `notify()` and I can propagate an event from a child element.  

Since I've resolved the problem I'll close this issue.  Overall I could probably have saved a week if there had been documentation on migrating custom components from NS 6 to 7+.
